include/envoy/stats/stats.h

```diff
@@ -20,39 +21,90 @@ class Instance;
 namespace Stats {
 
 /**
+ * General representation of a tag.
+ */
+struct Tag {
+  std::string name_;
+  std::string value_;
+};
+
+class TagExtractor {
+public:
+  virtual ~TagExtractor() {}
+
+  /**
+   * Identifier for this tag.
+   */
+  virtual std::string name() const PURE;
+
+  /**
+   * Updates the tag extracted name and the set of tags by extracting the tag represented by this
+   * TagExtractor. If the tag is not represented in the current tag_extracted_name, nothing will be
+   * modified.
+   * @param[in,out] tag_extracted_name name from which the tag will be removed if found to be
```

Comment:
Side comment: I don't think I've seen this `@param` syntax used before in Envoy, but as long as it's valid Doxygen then seems useful. I think to some extent this is implied by argument `const`ness already though.

Reply:
Just a note on that (may be irrelevant considering we're going to make this the return value).  This is only used if it's going to be both and in and out parameter rather than just an out parameter.  A lack of const only tells you that the object will be written to.  That's generally the default for non-const inputs (at least for simple stdlib containers and primitives).  For instance, I'd consider the vector in this case to be an out parameter because the function will act identically no matter what is currently in the vector.  However, the tag_extracted_name's current state will change the logical behavior of the function (whether a tag is actually extracted) and the state will be modified.